<a href="https://colab.research.google.com/github/vy-phung/genomic-data-science-project-about-fetus-and-adult/blob/main/Predicted_and_classified_genes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Prediction and Classification from fetal and adult genes 

## Objectives

I will use 3 types of genomic table data to identify if we can predict ages and classify age_group (fetus or adult) and genders of these 10 samples: 
1. Edata_tr.csv (15559 genes, haven't statistically analyzed but already filtered)
2. Data for regulated gene.csv (5566 genes, already statistically analyzed)
3. 2 most up-down regulated genes (2 genes, extracting from data for regulated gene.csv)

*Getting data in Github: genomic data (tidy_data folder); phenotype data (sample_data folder)*

**Steps:**
*   Classifying age_group (fetus or adult) of samples and gender (female or male) by using Logistic regression
*   Predicting age which is continuous data by using Multi linear regression, Ridge Regression with the support of Grid alpha search, and Random Forest Regressor


<h1>Table of content</h1>
<ul>
    <li><a href="#ref1">Part 1: Algorithm </a></li>
    <li><a href="#ref2">Part 2: Analyzing table type 1 </a></li>
    <li><a href="#ref3">Part 3: Analyzing table type 2 </a></li>
    <li><a href="#ref4">Part 4: Analyzing table type 3</a></li>



<h2>Part 1: Predicting and Classifying Algorithm</h2>

In [ ]:
#@title Import packages 
# Package
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Ridge
from sklearn.preprocessing import PolynomialFeatures

In [ ]:
#@title Logistic regression & Multi linear regression
class Prediction:
  def __init__(self,x,y,lr):
    self.x = x
    self.y = y
    self.lr = lr
  def train_test(self):
      # lr is regressor used to predict or classify 
    x_train, x_test, y_train, y_test = train_test_split(self.x, self.y, test_size=0.20, random_state=0)
    self.lr.fit(x_train,y_train)
    self.x_test = x_test
    self.x_train = x_train
    self.y_test = y_test
    self.y_train = y_train
  def predict(self,x,y): 
    self.train_test() 
    if str(x.values) == str(self.x.values) and str(y.values) == str(self.y.values):
      x = self.x_test
      y = self.y_test
      print("number of test samples :", self.x_test.shape[0])
      print("number of training samples:",self.x_train.shape[0])
    print("actual data: ",y.values)
    print("predicted data: ",self.lr.predict(x)) 
  def cnf(self,x,y):
    if y.iloc[:,0].dtype == 'O': # logistic regression for binary variable
      self.train_test()
      if str(x.values) == str(self.x.values) and str(y.values) == str(self.y.values):
        self.x == x
        self.y == y 
      print("confusion matrix:",metrics.confusion_matrix(y, self.lr.predict(x)))
  def score(self,x,y):
    if y.iloc[:,0].dtype == 'float64': # regression for continuous data 
      self.train_test()
      if str(x.values) == str(self.x.values) and str(y.values) == str(self.y.values):
        print('The R-square of testing data: ', self.lr.score(self.x_test, self.y_test))
        print('The R-square of training data: ', self.lr.score(self.x_train, self.y_train))
        print('The mean square error of actual and predicted fitting data is: ', \
          mean_squared_error(self.y, self.lr.predict(self.x)))
        return
      print('The R-square: ', self.lr.score(x, y))

In [ ]:
#@title Ridge regression and Grid Search
class other_prediction:
  def __init__(self,x,y,parameters):
    self.x = x
    self.y = y
    self.parameters = parameters
  def train_test(self):
      # lr is regressor used to predict or classify 
    x_train, x_test, y_train, y_test = train_test_split(self.x, self.y, test_size=0.20, random_state=0)
    RR=Ridge()
    self.Grid = GridSearchCV(RR, self.parameters,cv=4)
    self.x_test = x_test
    self.x_train = x_train
    self.y_test = y_test
    self.y_train = y_train
  def Ridge(self,x,y):
    if y.iloc[:,0].dtype == 'float64':
      self.train_test()
      self.Grid.fit(self.x, self.y)
      BestRR=self.Grid.best_estimator_
      RigeModel=Ridge(alpha=BestRR.alpha)
      RigeModel.fit(self.x, self.y)
      print(BestRR)
      if str(x.values) == str(self.x.values) and str(y.values) == str(self.y.values):
        x = self.x_test
        y = self.y_test
      print("R^2 of data:",RigeModel.score(x, y))
      print("predict data:",RigeModel.predict(x))
      print("actual testing data:",y.values)

<h2>Part 2: Table type 1</h2>

#### <b>Data processing<b/>

In [ ]:
edata = pd.read_csv("edata_tr.csv")
edata

Unnamed: 0  SRR1554534  SRR1554535  ...  SRR1554539  SRR1554538  SRR1554537
0            A1BG    9.143633    8.583227  ...    8.429514    7.632873    8.489938
1             ADA    8.646385    9.145974  ...    8.253862    7.831596    7.197164
2            CDH2   13.562468   13.585906  ...   13.717079   14.708753   15.051336
3            AKT3   13.479431   13.159273  ...   14.383520   14.749481   15.584859
4      ZBTB11-AS1    7.746129    7.539217  ...    7.654166    7.625887    7.705363
...           ...         ...         ...  ...         ...         ...         ...
15554        CDH1    9.283095    9.269662  ...    8.933461    9.185240    9.167789
15555     SLC12A6   13.031252   12.899341  ...   13.345966   13.644767   13.585326
15556       PTBP3   11.367093   11.867973  ...   12.386899   12.840028   12.564455
15557       DGCR2   15.038492   14.455368  ...   14.553994   13.960779   14.168734
15558    CASP8AP2   11.208092   11.459779  ...   11.802491   12.758373   12.705135

[15559 rows x 11 columns]

In [ ]:
edata = edata.set_index("Unnamed: 0")
edata.index.names  = [None]
edata_T = edata.T
edata_T

A1BG       ADA       CDH2  ...      PTBP3      DGCR2   CASP8AP2
SRR1554534  9.143633  8.646385  13.562468  ...  11.367093  15.038492  11.208092
SRR1554535  8.583227  9.145974  13.585906  ...  11.867973  14.455368  11.459779
SRR1554568  8.105891  7.303994  14.957846  ...  13.315986  13.976339  12.916206
SRR1554561  9.206773  8.001912  13.652418  ...  11.766260  15.153004  11.188994
SRR1554567  6.995598  7.351627  14.829826  ...  12.716064  13.996213  12.598715
SRR1554536  8.762864  9.770115  13.812075  ...  11.924352  14.484859  11.474929
SRR1554541  8.373665  7.843839  14.545283  ...  12.990262  13.867110  12.611138
SRR1554539  8.429514  8.253862  13.717079  ...  12.386899  14.553994  11.802491
SRR1554538  7.632873  7.831596  14.708753  ...  12.840028  13.960779  12.758373
SRR1554537  8.489938  7.197164  15.051336  ...  12.564455  14.168734  12.705135

[10 rows x 15559 columns]

In [ ]:
pheno = pd.read_csv("/content/pheno_sample.csv")
pheno = pheno.set_index("Unnamed: 0")
pheno.index.names = [None]
pheno

fetus_adult  age_group      age     sex  gender
SRR1554534       adult          1  40.4200    male       1
SRR1554535       adult          1  41.5800    male       1
SRR1554568       fetus          0  -0.4986    male       1
SRR1554561       adult          1  43.8800    male       1
SRR1554567       fetus          0  -0.4027    male       1
SRR1554536       adult          1  44.1700  female       0
SRR1554541       fetus          0  -0.3836    male       1
SRR1554539       adult          1  36.5000  female       0
SRR1554538       fetus          0  -0.4027  female       0
SRR1554537       fetus          0  -0.3836  female       0

In [ ]:
x = edata_T

#### <b>Logistic regression<b/>

**Classifying fetus or adult based on age_group**

In [ ]:
y = pheno[["age_group"]]
y

age_group
SRR1554534          1
SRR1554535          1
SRR1554568          0
SRR1554561          1
SRR1554567          0
SRR1554536          1
SRR1554541          0
SRR1554539          1
SRR1554538          0
SRR1554537          0

In [ ]:
lr = LogisticRegression()
p = Prediction(x,y,lr)
p.predict(x,y)
p.cnf(x,y)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:985: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


number of test samples : 2
number of training samples: 8
actual data:  [[0]
 [0]]
predicted data:  [0 0]


**Classifying gender of samples**

In [ ]:
y_gender = pheno[["gender"]]
y_gender

gender
SRR1554534       1
SRR1554535       1
SRR1554568       1
SRR1554561       1
SRR1554567       1
SRR1554536       0
SRR1554541       1
SRR1554539       0
SRR1554538       0
SRR1554537       0

In [ ]:
p = Prediction(x,y_gender,lr)
p.predict(x,y_gender)
p.cnf(x,y_gender)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:985: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


number of test samples : 2
number of training samples: 8
actual data:  [[1]
 [0]]
predicted data:  [0 0]


#### <b>Predicting age (continuous data)<b/>

In [ ]:
y_age = pheno[["age"]]
y_age

age
SRR1554534  40.4200
SRR1554535  41.5800
SRR1554568  -0.4986
SRR1554561  43.8800
SRR1554567  -0.4027
SRR1554536  44.1700
SRR1554541  -0.3836
SRR1554539  36.5000
SRR1554538  -0.4027
SRR1554537  -0.3836

**Multi linear regression**

In [ ]:
lr = LinearRegression()
p = Prediction(x,y_age,lr)
p.predict(x,y_age)
p.score(x,y_age)

number of test samples : 2
number of training samples: 8
actual data:  [[-0.4986]
 [-0.4027]]
predicted data:  [[-3.38719242]
 [-1.75369115]]
The R-square of testing data:  -2210.4501135869386
The R-square of training data:  1.0
The mean square error of actual and predicted fitting data is:  1.0169143259568714


The R-square of testing data:  -2210.4501135869386

The R-square of training data:  1.0

We can see that R^2 of testing data is negatively large, while that of training data is fit to statistical model

=> It is overfitting

**Using grid search to find suitable alpha for ridge model**

In [ ]:
parameters= [{'alpha': [0.001,0.1,1, 10, 100, 1000, 10000, 100000, 100000]}]
o = other_prediction(x,y_age,parameters)
o.Ridge(x,y_age)

Ridge(alpha=1000)
R^2 of data: -31.726080041146417
predict data: [[-0.88566156]
 [-0.428606  ]]
actual testing data: [[-0.4986]
 [-0.4027]]


R^2 of data: -31.726080041146417

We can see that using Ridge model combined with alpha 1000 can increase R^2
 
=> It might be better to use than multi linear regression model.


<h2>Part 3: Table type 2</h2>

#### <b>Data processing<b/>

In [ ]:
up_down = pd.read_csv("data for regulated gene.csv")
up_down

Unnamed: 0      baseMean  ...  SRR1554538  SRR1554537
0          A1BG    380.846902  ...    7.632873    8.489938
1           A2M  13185.383185  ...   12.930901   12.535831
2         A2ML1    484.632776  ...    7.500564    7.351208
3        A4GALT    412.311453  ...    7.292369    6.723675
4          AARD    124.885643  ...    5.264764    5.737391
...         ...           ...  ...         ...         ...
7026       ZXDA   1338.880407  ...   10.778896   10.978820
7027       ZXDB   2840.638643  ...   11.716040   11.857990
7028       ZXDC  14083.362352  ...   14.424315   14.336139
7029        ZYX  14229.672689  ...   13.208680   13.456611
7030       ZZZ3   5105.819591  ...   12.714798   12.679419

[7031 rows x 17 columns]

In [ ]:
select = ["Unnamed: 0","SRR1554534",	"SRR1554535",	"SRR1554568",	"SRR1554561",	"SRR1554567",	"SRR1554536",	"SRR1554541",	"SRR1554539",	"SRR1554538",	"SRR1554537"]
up_down1 = up_down[select]
up_down1 = up_down1.set_index("Unnamed: 0")
up_down1.index.names = [None]
up_down_T = up_down1.T
up_down_T

A1BG        A2M      A2ML1  ...       ZXDC        ZYX       ZZZ3
SRR1554534  9.143633  13.657416   8.848175  ...  12.904692  14.633231  11.142157
SRR1554535  8.583227  13.795969   9.062477  ...  12.734604  13.965675  11.893657
SRR1554568  8.105891  12.584194   6.909735  ...  14.242381  13.345845  12.719315
SRR1554561  9.206773  13.003245   8.644313  ...  12.935801  14.427006  11.499921
SRR1554567  6.995598  12.597035   7.457088  ...  14.026589  13.168138  12.594819
SRR1554536  8.762864  15.235020  10.622862  ...  12.953526  13.685386  11.839166
SRR1554541  8.373665  12.938130   6.885452  ...  14.199765  13.205510  12.744916
SRR1554539  8.429514  13.881265   7.708383  ...  12.951150  13.632358  12.019470
SRR1554538  7.632873  12.930901   7.500564  ...  14.424315  13.208680  12.714798
SRR1554537  8.489938  12.535831   7.351208  ...  14.336139  13.456611  12.679419

[10 rows x 7031 columns]

In [ ]:
x1 = up_down_T

#### <b>Logistic regression<b/>

**Classifying fetus or adult based on age_group**

In [ ]:
lr = LogisticRegression()
p = Prediction(x1,y,lr)
p.predict(x1,y)
p.cnf(x1,y)

number of test samples : 2
number of training samples: 8
actual data:  [[0]
 [0]]
predicted data:  [0 0]


/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:985: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


**Classifying gender of samples**

In [ ]:
p = Prediction(x1,y_gender,lr)
p.predict(x1,y_gender)
p.cnf(x1,y_gender)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:985: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


number of test samples : 2
number of training samples: 8
actual data:  [[1]
 [0]]
predicted data:  [0 1]


#### <b>Predicting age (continuous data)<b/>

**Multi linear regression**

In [ ]:
lr = LinearRegression()
p = Prediction(x1,y_age,lr)
p.predict(x1,y_age)
p.score(x1,y_age)

number of test samples : 2
number of training samples: 8
actual data:  [[-0.4986]
 [-0.4027]]
predicted data:  [[-3.72575422]
 [-1.9660493 ]]
The R-square of testing data:  -2795.314245992872
The R-square of training data:  1.0
The mean square error of actual and predicted fitting data is:  1.2858585410344976


The R-square of testing data:  -2759.9297929965064

The R-square of training data:  1.0

We can see that R^2 of testing data is negatively large, while that of training data is exactly fit which also means statistical model fits

It is overfitting

**Using grid search to find suitable alpha for ridge model to predict age of sample**

In [ ]:
parameters= [{'alpha': [0.001,0.1,1, 10, 100, 1000, 10000, 100000, 100000]}]
o = other_prediction(x1,y_age,parameters)
o.Ridge(x1,y_age)

Ridge(alpha=1000)
R^2 of data: -94.94043343888032
predict data: [[-1.15500159]
 [-0.50423776]]
actual testing data: [[-0.4986]
 [-0.4027]]


R^2 of data: -94.94043343888032

We can see that using Ridge model combined with alpha 1000 can increase the R^2, and reduce overfitting compared with multi linear regression 

<h2>Part 4: Table type 3</h2>

#### <b>Data processing <b/>

In [ ]:
# The most up-regulated genes
up = up_down[up_down['log2FoldChange'] > 1]
up = up[up["padj"] == min(up["padj"])]
up = up.set_index("Unnamed: 0")
up.index.names = [None]
up

baseMean  log2FoldChange  ...  SRR1554538  SRR1554537
ST8SIA2  21470.98555         7.44298  ...   14.917903   15.015816

[1 rows x 16 columns]

In [ ]:
# The most down-regulated genes
down = up_down[up_down['log2FoldChange'] < -1]
down = down[down["padj"] == min(down["padj"])]
down = down.set_index("Unnamed: 0")
down.index.names = [None]
down

baseMean  log2FoldChange  ...  SRR1554538  SRR1554537
BCL2L2  21968.220153       -2.676407  ...   12.718466   12.793366

[1 rows x 16 columns]

In [ ]:
up1 = up[["SRR1554534",	"SRR1554535",	"SRR1554568",	"SRR1554561",	"SRR1554567",	"SRR1554536",	"SRR1554541",	"SRR1554539",	"SRR1554538",	"SRR1554537"]]
up1_T = up1.T
up1_T

ST8SIA2
SRR1554534   8.614598
SRR1554535   8.485263
SRR1554568  14.645985
SRR1554561   8.707785
SRR1554567  14.756824
SRR1554536   8.396410
SRR1554541  15.159507
SRR1554539   8.971893
SRR1554538  14.917903
SRR1554537  15.015816

In [ ]:
down1 = down[["SRR1554534",	"SRR1554535",	"SRR1554568",	"SRR1554561",	"SRR1554567",	"SRR1554536",	"SRR1554541",	"SRR1554539",	"SRR1554538",	"SRR1554537"]]
down1_T = down1.T
down1_T

BCL2L2
SRR1554534  14.967316
SRR1554535  14.943844
SRR1554568  12.776112
SRR1554561  15.151775
SRR1554567  12.714420
SRR1554536  14.999478
SRR1554541  12.761278
SRR1554539  15.053436
SRR1554538  12.718466
SRR1554537  12.793366

In [ ]:
x2 = pd.concat([up1_T,down1_T],axis=1)
x2

ST8SIA2     BCL2L2
SRR1554534   8.614598  14.967316
SRR1554535   8.485263  14.943844
SRR1554568  14.645985  12.776112
SRR1554561   8.707785  15.151775
SRR1554567  14.756824  12.714420
SRR1554536   8.396410  14.999478
SRR1554541  15.159507  12.761278
SRR1554539   8.971893  15.053436
SRR1554538  14.917903  12.718466
SRR1554537  15.015816  12.793366

#### <b>Logistic regression<b/>

**Classifying by age_group**

In [ ]:
lr = LogisticRegression()
p = Prediction(x2,y,lr)
p.predict(x2,y)
p.cnf(x2,y)

number of test samples : 2
number of training samples: 8
actual data:  [[0]
 [0]]
predicted data:  [0 0]


/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:985: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


**Classifying by gender**

In [ ]:
# gender
p = Prediction(x2,y_gender,lr)
p.predict(x2,y_gender)
p.cnf(x2,y_gender)

number of test samples : 2
number of training samples: 8
actual data:  [[1]
 [0]]
predicted data:  [1 1]


/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:985: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


#### <b>Predicting age (continuous data)<b/>

**Multi linear regression**

In [ ]:
lr = LinearRegression()
p = Prediction(x2,y_age,lr)
p.predict(x2,y_age)
p.score(x2,y_age)

number of test samples : 2
number of training samples: 8
actual data:  [[-0.4986]
 [-0.4027]]
predicted data:  [[ 1.38679745]
 [-0.25893323]]
The R-square of testing data:  -776.5288225177012
The R-square of training data:  0.9934719800426676
The mean square error of actual and predicted fitting data is:  2.511263182559502


The R-square of testing data:  -776.5288225177012

We can see that R^2 of testing data is negatively large, while that of training data nearly fits statistical model

=> It is overfitting

**Using grid search to find suitable alpha for ridge model**

In [ ]:
parameters= [{'alpha': [0.001,0.1,1, 10, 100, 1000, 10000, 100000, 100000]}]
o = other_prediction(x2,y_age,parameters)
o.Ridge(x2,y_age)

Ridge(alpha=1)
R^2 of data: -684.0498220950029
predict data: [[ 1.27614849]
 [-0.42280739]]
actual testing data: [[-0.4986]
 [-0.4027]]


R^2 of data: -684.0498220950029

**Random Forest Regression**

In [ ]:
from sklearn.ensemble import RandomForestRegressor

In [ ]:
lr = RandomForestRegressor(n_estimators=1)
p = Prediction(x2,y_age,lr)
p.predict(x2,y_age)
p.score(x2,y_age)

number of test samples : 2
number of training samples: 8
actual data:  [[-0.4986]
 [-0.4027]]
predicted data:  [-0.3836 -0.3836]
The R-square of testing data:  -1.8759972207754645
The R-square of training data:  0.9830837747102115
The mean square error of actual and predicted fitting data is:  5.582322500000003


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  # Remove the CWD from sys.path while we load stuff.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  # Remove the CWD from sys.path while we load stuff.


Using random forest regression with estimators = 1, we have R-square of testing data:  -1.8759972207754645 

In this table type 3, the random forest regression is better than multi linear regression and ridge model. This is because its R^2 of testing data is the highest, but it's still negative.  